golem-rust/src/transaction.rs

```diff
@@ -0,0 +1,167 @@
+// Copyright 2024 Golem Cloud
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::bindings::golem::api::host::{get_oplog_index, set_oplog_index, OplogIndex};
+use crate::mark_atomic_operation;
+use std::rc::Rc;
+
+/// Represents an atomic operation of the transaction which has a rollback action.
+///
+/// Implement this trait and use it within a `transaction` block.
+/// Operations can also be constructed from closures using `operation`.
+pub trait Operation<In, Out, Err>: Clone {
```

Comment:
This is a very data-oriented API, which makes it more difficult for the user to interleave other things, such as logging, whose behavior does not need to be compensated in the event of a rollback.

I wonder if we could achieve a less-invasive API by executing the actions immediately, and only storing the rollbacks.

Let's imagine two possible APIs:

```rust
transaction(move |tx| {
  // action is executed now, but compensation is stored in tx:
  tx.execute(|| { ... }, || { ... });
});
```

Or:

```rust
let txn = Transaction::new();

txn.begin_op();

// Arbitrary rust code:
do_something();

txn.end_op(|| { my_compensation(); });
```

With type classes and macros, we could improve upon this a lot:

```rust
pub trait Operation {
  type Out;
  type Err;

  fn execute(&self) -> Result<Out, Err>;
  fn compensate(&self, out: Out) -> Result<(), Err>;
}

#[golem_operation]
#[golem_compensation(my_compensation)]
pub fn my_operation() -> Result<bool, Error> {
}

pub fn my_compensation(b: bool) -> Result<(), Error> {
}

...
log("some logging stuff here");
txn.execute(my_operation);
log("some logging stuff there");
```

Reply:
You don't need to separate the operation definition from the execution in my version either:

```rust
tx.add(operation::<String, (), ()>(
            move |input: String| {
                log1.borrow_mut().push(format!("op1 execute {input}"));
                Ok(())
            },
            move |input: String| {
                log2.borrow_mut().push(format!("op1 rollback {input}"));
            },
));
```
Of course it can be simplified like the `execute` above, but we can only do that with this version and not my proposed "advanced one" (in #10)

The `begin_op`/`end_op` version cannot catch the errors happening in between, so there is no way to perform the rollback.

The macro to generate `Operation` instances is probably a good idea (although a bit magical to me :) and seems easy to do.

Reply:
What I mean by separate is that you cannot put code around the execution of the operation, because you are building a data structure, which will be executed later. It would be nice to "execute now" the operation, and keep track of the compensation, because it allows arbitrary code before and after the compensated operation.

Reply:
Ah I think I just named it confusingly. My `add` immediately executes the operation, and you can definitely have code around it. That's the whole reason transaction get's a `f: impl FnOnce(&mut Transaction) -> Out` and not a list of operations.

Reply:
I did not know this. Maybe we can rename to `execute`?

Let's create a ticket for the macro for operation. I think it can become bearable syntax with the macro, added to top-level functions.

Reply:
Added a first version of the macro. 

I'm going to extend `Operation` and the macro as well to have a way to write compensations that require the output of the action - with the current one it only gets the input, which is the only thing we can do when we want to run compensations on non-domain failures too. But for the simpler saga pattern it is easier if you don't need to derive the compensation from the input (and probably it is not even possible in many cases) so I am going to support both.